### PR TITLE
[WIP] PR: Add multiline editing capability in the Editor

### LINKED ISSUE #2112: Add multiline editing to the Editor
_From [geoffrey...@gmail.com](https://code.google.com/u/106789452866884306879/) on 2015-01-13T04:28:43Z_

Is a feature of multi-editing / multi-selection planned?

Basically the idea is to be able to set multiple cursors to edit at different places of the code at the same time.

Link with the ability to select several occurrence of the same word, it can be used as a very quick "find and replace" for instance.

Here an example: http://blog.jetbrains.com/ruby/2014/03/whats-mining-multiple-cursors-and-selection/

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=2112_

### PR DESCRIPTION
**Pull request by jnsebgosselin**
## Tasks

- [ ] Any new code on this PR should be Python 3 only
- [ ] Create helpers on the extension manager for common tasks (alt, ctrl, keys)
- [ ] Simplify the key_press handler to use helper methods (copy, paster and other actions)
- [ ] Provide a way to query for the current shortcuts sequences used for common actions
- [ ] Check usage of Mac pressed keys. On mac leaving a key pressed for a while sometimes produces extra options (@steff456, @juanis2112)
- [ ] Check Alt+Gr behavior on windows since this is source of weirdness (@dalthviz)
- [ ] Check LSP calls are working as expected after a extension has accepted an event.
   - [ ] If this is not the case try to extend the Extension base class to do this handling. Extensions should not need to care about LSP issues in general, unless the extension is doing something with LSP results. In that case we will also need to expose LSP signals so that the extension editors can connect to them.
- [ ] Update documentation inside the extension editor to explain usage. The event.accept, ignore way of handling signals and events as well as the purpose of these signals.
- [ ] Check color of Cursor is correctly querying the editor color theme.
- [ ] Check the cursors are blinking in sync (this was one of the issues we had in the past.
- [ ] Check performance of editor if many cursors are added.
- [ ] Check how cursor history is affected by multiple cursors (cursor history should probably be disabled if multi line editing @ccordoba12 ?)
- [ ] We need test for all the corner cases :-) 

### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based your PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP8](https://www.python.org/dev/peps/pep-0008/) for code style
* [x] Ensured your pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [ ] Wrote at least one-line docstrings for any new functions
* [ ] Added at least one unit test covering the changes, if at all possible
* [x] Described your changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] Included a screenshot, if this PR makes any visible changes to the UI

## Description of Changes

The idea is to be able to set multiple cursors to edit at different places of the code at the same time like it is done in Sublime for example.

This feature is not supported by the Qt framework, so we need to go around it and re-implement a lot of low level stuff.

mulitline_spyder

### Issue(s) Resolved

Fixes #2112
Depends on PR #5002

### CODE REVIEW

**Comment by ccordoba12** [comment]
Thanks a lot @jnsebgosselin! This is looking great!

@rlaverde, please review this one and give @jnsebgosselin feedback about his implementation.

**Comment by goanpeca** [comment]
@jnsebgosselin great work :-)

Question, why did we need  to create a new class on top of the base editor?

Also, cursors should occupy a bit less than the full height of text so when pile together, they do not look like a single super long line

**Comment by rlaverde** [comment]
@jnsebgosselin  Thanks for your contribution, It looks pretty good

> Question, why did we need to create a new class on top of the base editor? 

I agree with @goanpeca, I think we could use mixing instead of adding a new layout of inheritance to the CodeEditor

A solution maybe out of the scope of this PR:
I like the idea of how pyqode split the functionality in different "Modes" (I'll call them something like editor features), they are plugable classes, like the panels, taht keep reference top the CodeEditor, and implement methods that are fired according to CodeEditor events

something like this:

```
class MultiCursorsEditorFeature(EditorFeature):
     def _on_key_pressed():
          ....
     def _on_key_released():

     def register_editor_feature():
           self._editor.sig_key_press.connect(self._on_key_pressed)
           self._editor.sig_key_press.connect(self._on_key_pressed)

class CodeEditor():

    def __init__(self):
         editor_features = EditorFeaturesManager
         editor_features.register(MultiCursorEditorsFeature)

    def KeyPressEvent(self, event):
         self.sig_key_press.emit(event)
```

You could take a look at some pyqode modes https://github.com/pyQode/pyqode.core/blob/master/pyqode/core/modes/autocomplete.py

We will need to create a `EditorFeaturesManager()` and `EditorFeatures()` (we already have pyqode Mode that we use for panels) I think we could reuse some panels logic and It wont be really hard to implement.

@ccordoba12 @goanpeca What do you think?

**Comment by goanpeca** [comment]
@rlaverde sounds good. But what to do in the meantime :-p

**Comment by rlaverde** [comment]
> But what to do in the meantime :-p

I don't know, putting more code in the already complex CodeEditor doesn't seem as a solution 😞

**Comment by ccordoba12** [comment]
> We will need to create a EditorFeaturesManager() and EditorFeatures() (we already have pyqode Mode that we use for panels) I think we could reuse some panels logic and It wont be really hard to implement.

I really like this idea! @rlaverde, could you implement this so that @jnsebgosselin can use it to implement this feature?

**Comment by rlaverde** [comment]
Which name should I use? Modes? EditorFeatures? any other suggestion?

**Comment by jnsebgosselin** [comment]
This looks neat! I like the way it is possible to easily turn off or on modes with that approach.

**Comment by ccordoba12** [comment]
> Which name should I use? Modes? EditorFeatures?

I think `Features` is enough. Since everything related to a plugin is going to live under the same directory after we merge `split-plugins`, I think there's no need to be more explicit.

**Comment by ccordoba12** [comment]
@jnsebgosselin, please rebase this PR and use @rlaverde's new Editor extensions to implement this feature.

**Comment by ccordoba12** [comment]
@jnsebgosselin, @rlaverde can finish this one. Do you agree?

**Comment by jnsebgosselin** [comment]
@ccordoba12 Sure thing.

**Comment by ccordoba12** [comment]
@rlaverde, please make this an Editor extension and add tests for it.

File: spyder/plugins/editor/extensions/multicursorseditor.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+"""This module contains the close quotes editor extension."""
```
Comment (ccordoba12) [inline spyder/plugins/editor/extensions/multicursorseditor.py]: This is wrong. Please also add an extra blank before this line.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 2]
```diff
@@ -63,13 +63,13 @@
...
 from spyder.plugins.editor.utils.folding import IndentFoldDetector
-from spyder.plugins.editor.utils.decoration import TextDecorationsManager
-from spyder.plugins.editor.extensions.manager import (
-    EditorExtensionsManager)
-from spyder.plugins.editor.extensions.closequotes import (
-    CloseQuotesExtension)
+from spyder.plugins.editor.extensions.manager import EditorExtensionsManager
+from spyder.plugins.editor.extensions.closequotes import CloseQuotesExtension
+from spyder.plugins.editor.extensions.multicursorseditor import (
+    MultiCursorsEditorExtension)
```
[the remark this answers is not in the document]
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/codeeditor.py]: Good idea, I will do this. 👍

[the remark this answers is not in the document]
Comment (ccordoba12) [reply spyder/plugins/editor/widgets/codeeditor.py]: Annd, EditorExtensionsManager could go there too.

File: spyder/plugins/editor/widgets/codeeditor.py [hunk 3]
```diff
@@ -81,7 +81,7 @@
 try:
     import nbformat as nbformat
     from nbconvert import PythonExporter as nbexporter
-except:
+except ImportError:
```
[the remark this answers is not in the document]
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/codeeditor.py]: Can we change `ImportError` by `Exception` or it is still better to keep it bare?

File: spyder/plugins/editor/widgets/codeeditor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/codeeditor.py]: Done. Is it ok if I did this in the init ?

```
from .manager import EditorExtensionsManager
from .closequotes import CloseQuotesExtension
from .multicursorseditor import MultiCursorsEditorExtension
```

[the remark this answers is not in the document]
Comment (jnsebgosselin) [reply spyder/plugins/editor/widgets/codeeditor.py]: ... and the answer is no.

**Comment by ccordoba12** [comment]
@jnsebgosselin, do you plan to come back to this one? It'b really nice to have it for beta3.

**Comment by jnsebgosselin** [comment]
Yes, I will resume work on this very soon.

**Comment by ccordoba12** [comment]
Nice!! Thanks a lot!

**Comment by goanpeca** [comment]
This work is pretty cool @jnsebgosselin we will squash the commits and rebase it with master so we can try to use this PR.

Will also add some tasks on the description as @dalthviz @juanis2112 and @steff456 will continue this work. Thanks a lot!

**Comment by jnsebgosselin** [comment]
> This work is pretty cool @jnsebgosselin, @dalthviz will squash the commits (to 1 to make it easy to rebase) and then rebase it with master so we can try to use this PR.
> 
> Will also add some tasks on the description as @dalthviz @juanis2112 and @steff456 will continue this work. Thanks a lot!

Oh great, I thought this would be useless, so I'm happy if this helps! I'm sorry I haven't had time to look at this to try to help you further but at this point, I did this so long ago that I don't remember anything lol

**Comment by goanpeca** [comment]
>Oh great, I thought this would be useless, so I'm happy if this helps! I'm sorry I haven't had time to look at this to try to help you further but at this point, I did this so long ago that I don't remember anything lol

What you did works just fine!, the extension logic on the editor (that we borrowed from `pyqode` project) has remain intact, so all the work is completely on track to be reused and completed.

I also forgot about this as well, but after reviewing your work again, it all made sense, so thanks again, and no worries :-p

**Comment by brupelo** [comment]
@jnsebgosselin I must to say I'm impressed, really promising PR/WIP you've created over here! one question, when undoing/redoing... will the multiple selections be preserved? Please take a read to see what I mean by that in this thread https://bugreports.qt.io/browse/QTCREATORBUG-16013 . I'm asking this cos one of the main weak points of Scintilla for instance is the fact that the different selections will be gone when undoing/redoing... But in proper implementations such as SublimeText, Codemirror, Monarch, etc... such selections will be preserved.

And just for the record, when I say I'm impressed I really mean it... you won't see this "must to have on a modern text editor" feature on almost any existing Qt widgets out there on all github... so... hats off :) . Pasting some references I've found over time regarding to this subject in case it helps:

Text editors
------------

https://github.com/ajaxorg/ace
https://github.com/andreikop/qutepart
https://github.com/awm/synhtor/wiki/Research
https://github.com/codemirror/CodeMirror
https://github.com/edbee/edbee
https://github.com/edbee/edbee-app
https://github.com/edbee/edbee-data
https://github.com/edbee/edbee-examples
https://github.com/edbee/edbee-lib
https://github.com/githole/Live-Coder
https://github.com/howl-editor
https://github.com/hydrargyrum/eye
https://github.com/JuBan1/notepadqq
https://github.com/JuBan1/OpenTextEdit
https://github.com/kai66673/PythonEditor
https://github.com/luchko/QCodeEditor
https://github.com/lukedan/codepad
https://github.com/martinrotter/textosaurus
https://github.com/matkuki/ExCo
https://github.com/prymatex/prymatex
https://github.com/pybee/seasnake
https://github.com/pyQode/pyqode.core
https://github.com/richrd/suplemon
https://github.com/SergeySatskiy/codimension
https://github.com/smathot/QProgEdit
https://github.com/spyder-ide/spyder
https://github.com/trishume/syntect
https://github.com/tsujan/FeatherPad
https://github.com/waddlesplash/Heidi
https://kate-editor.org/
https://sourceforge.net/projects/synwrite

Text rendering
--------------
https://learnopengl.com/In-Practice/Text-Rendering
https://wdobbie.com/post/gpu-text-rendering-with-vector-textures/
https://medium.com/@evanwallace/easy-scalable-text-rendering-on-the-gpu-c3f4d782c5ac
http://jcgt.org/published/0006/02/02/
https://blog.mapbox.com/drawing-text-with-signed-distance-fields-in-mapbox-gl-b0933af6f817
https://aras-p.info/blog/2017/02/15/Font-Rendering-is-Getting-Interesting/
https://medium.com/@calebfaith/implementing-msdf-font-in-opengl-ea09a9ab7e00
https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch25.html
https://www.microsoft.com/en-us/research/wp-content/uploads/2005/01/p1000-loop.pdf
https://wildfiregames.com/forum/index.php?/topic/17365-freetype-ttf-fonts-in-opengl/

https://github.com/rougier/freetype-gl
https://github.com/servo/pathfinder
https://github.com/Chlumsky/msdfgen
https://github.com/behdad/glyphy
https://github.com/0ad/0ad
----------------------------------------

Your implementation reminds me to `https://github.com/JuBan1/OpenTextEdit` ... cool stuff :D

**Comment by ccordoba12** [comment]
Closing as superseded by #22996. Thanks for your initial work on this @jnsebgosselin!
